app/models/user.rb

```diff
@@ -19,4 +19,15 @@ class User < ApplicationRecord
   def downcase_email
     email.downcase!
   end
+
+  class << self
+    private
+    def digest string
+      cost = if ActiveModel::SecurePassword.min_cost
+               BCrypt::Engine::MIN_COST
+             else BCrypt::Engine.cost
```

Comment:
cho này là sao nhỉ?

Reply:
Chỗ đó là password digest được khởi tạo bằng cách sử dụng brypt (thông qua has_secure_password). string là một hashes, cost là tham số chi phí xác định chi phí tính toán để tính hash. Sử dụng chi phí cao khiến cho tính toán khó có thể sử dụng hàm băm để xác định mật khẩu ban đầu, đây là một biện pháp phòng ngừa an ninh quan trọng trong môi trường production